File: src/advent/2023/day8.clj
(ns advent.2023.day8
  "Advent of Code 2023, day 8: Haunted Wasteland"
  (:require [advent.helpers :as h]
            [clojure.string :as str]))

(def puzzle-input (h/slurp-resource "2023/day8.txt" h/slurp-lines))

(defn mk-graph [nodes]
  (reduce (fn [acc [a l r]] (assoc acc a [l r])) {}
          (map (fn [node] (remove str/blank? (str/split node  #"[^\w]"))) nodes)))

(defn get-node [graph c d] (get-in graph [c (if (= d \L) 0 1)]))

;;; Part 1

(defn run-graph [graph ds]
  (loop [ds (cycle ds) i 0 c "AAA"]
    (if (= "ZZZ" c)
      i
      (recur (rest ds) (inc i) (get-node graph c (first ds))))))

(defn puzzle1 [[ds _ & nodes]]
  (run-graph (mk-graph nodes) ds))

;;; Part 2
;;; Finding whenever each individual path halts & returning
;;; lowest common multiplier of all halt indexes.

(defn gcd [x y]
  (if (zero? y)
    x
    (gcd y (mod x y))))

(defn lcm [x y]
  (/ (* x y) (gcd x y)))

(defn run-graph2 [graph dirs]
  (loop [dirs (cycle dirs)
         i 0
         found-lcm 1
         c (filter #(str/ends-with? % "A") (keys graph))]
    (if (seq c)
      (recur (rest dirs)
             (inc i)
             (if (some #(str/ends-with? % "Z") c) (lcm found-lcm i) found-lcm)
             (->> (remove #(str/ends-with? % "Z") c)
                  (map #(get-node graph % (first dirs)))))
      found-lcm)))

(defn puzzle2 [[dirs _ & nodes]]
  (run-graph2 (mk-graph nodes) dirs))

File: src/rust/year2023/day8.rs
// Advent of Code 2023, day 8: Haunted Wasteland
// https://adventofcode.com/2023/day/8

use num::integer::lcm;
use std::collections::HashMap;

fn puzzle1(dirs: &str, grid: HashMap<&str, (&str, &str)>) -> usize {
    let mut i = 0;
    let mut di = dirs.chars().cycle();
    let mut f = "AAA";
    loop {
        if f == "ZZZ" {
            return i;
        } else {
            i += 1;
            if di.next().unwrap() == 'L' {
                f = grid.get(f).unwrap().0;
            } else {
                f = grid.get(f).unwrap().1;
            }
        }
    }
}

fn puzzle2(dirs: &str, grid: HashMap<&str, (&str, &str)>) -> usize {
    let mut i = 0;
    let mut m = 1;
    let mut di = dirs.chars().cycle();
    let mut fs: Vec<&str> = grid.keys().filter(|x| x.ends_with('A')).copied().collect();

    let mut rem: Option<usize> = None;

    loop {
        if fs.is_empty() {
            return m;
        }

        let d = di.next().unwrap();

        for j in 0..fs.len() {
            if fs[j].ends_with('Z') {
                m = lcm(m, i);
                rem = Some(j);
            } else {
                let f = grid.get(fs[j]).unwrap();
                if d == 'L' {
                    fs[j] = f.0;
                } else {
                    fs[j] = f.1;
                }
            }
        }
        i += 1;

        if let Some(j) = rem {
            fs.remove(j);
            rem = None;
        }
    }
}

pub fn run(input: &str) {
    let mut i = input.split('\n');
    let dirs = i.next().unwrap();

    let graph: HashMap<&str, (&str, &str)> = i
        .skip(1)
        .map(|a| (&a[0..3], (&a[7..10], &a[12..15])))
        .collect();

    println!("{:?}", puzzle1(dirs, graph.clone()));
    println!("{:?}", puzzle2(dirs, graph.clone()));
}
